### PR TITLE
Generate: new length penalty docstring

### LINKED ISSUE #18208: length_penalty behavior is inconsistent with documentation
### System Info

- `transformers` version: 4.20.1
- Platform: Linux-5.4.120+-x86_64-with-glibc2.27
- Python version: 3.9.12
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.12.0+cu116 (True)
- Tensorflow version (GPU?): 2.9.1 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?:  yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@patrickvonplaten

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

`length_penalty` in language generation has different effects on the the length of the generation. Sometimes it makes the generation longer, sometimes it makes it shorter. This is very confusing as it is different from what the documentation says. Two previous issues touch on this problem: #4915 #16930

In Bart CNN/DM `length_penalty` **lengthens** the output.
```python
from transformers import pipeline
summarizer = pipeline("summarization", model='facebook/bart-large-cnn')
ARTICLE = """ New York (CNN)When Liana Barrientos was 23 years old, she got married in Westchester County, New York.
A year later, she got married again in Westchester County, but to a different man and without divorcing her first husband.
Only 18 days after that marriage, she got hitched yet again. Then, Barrientos declared "I do" five more times, sometimes only within two weeks of each other.
In 2010, she married once more, this time in the Bronx. In an application for a marriage license, she stated it was her "first and only" marriage.
Barrientos, now 39, is facing two criminal counts of "offering a false instrument for filing in the first degree," referring to her false statements on the
2010 marriage license application, according to court documents.
Prosecutors said the marriages were part of an immigration scam.
On Friday, she pleaded not guilty at State Supreme Court in the Bronx, according to her attorney, Christopher Wright, who declined to comment further.
After leaving court, Barrientos was arrested and charged with theft of service and criminal trespass for allegedly sneaking into the New York subway through an emergency exit, said Detective
Annette Markowski, a police spokeswoman. In total, Barrientos has been married 10 times, with nine of her marriages occurring between 1999 and 2002.
All occurred either in Westchester County, Long Island, New Jersey or the Bronx. She is believed to still be married to four men, and at one time, she was married to eight men at once, prosecutors say.
Prosecutors said the immigration scam involved some of her husbands, who filed for permanent residence status shortly after the marriages.
Any divorces happened only after such filings were approved. It was unclear whether any of the men will be prosecuted.
The case was referred to the Bronx District Attorney\'s Office by Immigration and Customs Enforcement and the Department of Homeland Security\'s
Investigation Division. Seven of the men are from so-called "red-flagged" countries, including Egypt, Turkey, Georgia, Pakistan and Mali.
Her eighth husband, Rashid Rajput, was deported in 2006 to his native Pakistan after an investigation by the Joint Terrorism Task Force.
If convicted, Barrientos faces up to four years in prison.  Her next court appearance is scheduled for May 18.
"""
print(summarizer(ARTICLE, max_length=512, min_length=30, do_sample=False, length_penalty=1))
print(summarizer(ARTICLE, max_length=512, min_length=30, do_sample=False, length_penalty=2))
```
Output:
`[{'summary_text': 'Liana Barrientos has been married 10 times, with nine of her marriages occurring between 1999 and 2002. She is believed to still be married to four men, and at one time, she was married to eight men at once.'}]`

`[{'summary_text': 'Liana Barrientos, 39, is charged with two counts of "offering a false instrument for filing in the first degree" In total, she has been married 10 times, with nine of her marriages occurring between 1999 and 2002. She is believed to still be married to four men.'}]`

In GPT-2 increasing `length_penalty` **shortens** the output.

```python
from transformers import pipeline
generator = pipeline('text-generation', model='gpt2', device=5)
print(generator("The White man worked as a", max_length=512, length_penalty=1))
print(generator("The White man worked as a", max_length=512, length_penalty=2))
```

Output:
`[{'generated_text': 'The White man worked as a receptionist for the British Consulate in Cairo and returned to Alexandria, where he was promoted to a military officer in 1953; in 1960 he worked as a consular officer, serving as secretary of state to President John F. Kennedy, and as a consul. In a conversation last fall, his grandfather told his sister Catherine, "We are going to make sure you are well."\n\nThe family is now living in a modest apartment, in a small part of town in the suburb of Alexandria.\n\n"We love you, and we love you," Catherine said, before she walked the five miles to the airport, where her husband, the first Egyptian president, has a $1 million plane ticket. The couple are still in touch with their three children, and will visit one next week.\n\nIn addition to the family, there are three other family members, one of whom has spent years as a caretaker for the hospital, which was the site of the largest civil conflict ever seen in modern Egypt. One was a nurse and family friend, who was paralyzed in a July 1975 accident.\n\n"It\'s just unbelievable," he told a reporter.\n\nThe funeral for one of the women who took her life last summer was held Wednesday at a church in the town of Dikun.\n\nIn his own words, the young woman\'s death marks a departure from his life.\n\n"I don\'t know if people would say I\'m the most important person in the world: I\'m the most beautiful person," he said. "But I did, but I will never forget that."'}]`

`[{'generated_text': "The White man worked as a mechanic.\n\nHe is said to have been very close with the White man's wife and three children. Other information came through during the early years of the investigation.\n\nPolice said they had asked the man to tell his story to police in order to gain information related to the white man's death.\n\nA source close to the father said the motive for the killings is still being investigated and the suspect was not a white man."}]`

### Expected behavior

Effect of `length_penalty` to be consistent with [documentation](https://huggingface.co/docs/transformers/v4.20.1/en/main_classes/text_generation#transformers.generation_utils.GenerationMixin.generate.length_penalty).

Currently the documentation says: 
"Exponential penalty to the length. 1.0 means that the beam score is penalized by the sequence length. 0.0 means no penalty. Set to values < 0.0 in order to encourage the model to generate longer sequences, to a value > 0.0 in order to encourage the model to produce shorter sequences."

### LINKED ISSUE #18971: generate() - documentation of `length_penalty' is misleading (and actually wrong)
### System Info

- `transformers` version: 4.20.1
- Platform: Linux-5.15.0-1014-azure-x86_64-with-glibc2.31
- Python version: 3.10.4
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.12.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@patrickvonplaten @Narsil  @gante

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

----

### Expected behavior

According to the documentation of the `generate()` function (transformers/generation_utils.py), the description of the `length_penalty` is as dollows:

>             length_penalty (`float`, *optional*, defaults to 1.0):
>                  Exponential penalty to the length. 1.0 means that the beam score is penalized by the sequence length.
>                  0.0 means no penalty. Set to values < 0.0 in order to encourage the model to generate longer
>                  sequences, to a value > 0.0 in order to encourage the model to produce shorter sequences.

However, this documentation is not aligned with the implementation of the `length_penalty` in methods which use it (like, `BeamHypotheses.add()`) or with the documentation of these methods:

Implementation:

`score = sum_logprobs / (hyp.shape[-1] ** self.length_penalty)`
**Note: the sum_logprobs is NEGATIVE(!!!), thus dividing it by a larger number making the score bigger**.

Documentation:

> length_penalty (`float`, *optional*, defaults to 1.0):
>             Exponential penalty to the length. 1.0 means no penalty. Set to values < 1.0 in order to encourage the
>             model to generate shorter sequences, to a value > 1.0 in order to encourage the model to produce longer
>             sequences.

I think the documentation of `BeamHypotheses.add()` is more correct and less misleading. I do understand that the sum_logprobs is the right logprob that represents the sequence logprobs, however, since the common practice for generation is to use the *mean* logprob: ` sum_logprobs / hyp.shape[-1]`, writing "1.0 means that the beam score is penalized by the sequence length" is misleading. Moreover, "Set to values < 0.0 in order to encourage the model to generate longer sequences, to a value > 0.0 in order to encourage the model to produce shorter sequences." is just not correct, since the sum_logprobs is NEGATIVE(!!!), thus dividing it by a larger number making the score bigger (bigger length_penalty --> bigger denominator --> bigger (negative) score (closer to zero). The documentation should be changed or explained more carefully (and be aligned with the implementations). 

I would change the documentation to something like:

> length_penalty (`float`, *optional*, defaults to 1.0):
>             Exponential penalty to the length. **0.0 means no penalty. 1.0 means the score of each sequence is the
>             log probability divided by the sequence length (the mean log probability, which is the common practice).**
>             Set to values < 1.0 in order to encourage the model to generate shorter sequences,
>             to a value > 1.0 in order to encourage the model to produce longer sequences.

Thanks

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes #18971 
Fixes #18208

The docstring for `length_penalty` was incomplete and incorrect -- it only has effect with beam-based strategies and the impact of this `float` argument is the other way around (> 0.0 promotes longer sequences, not shorter sequences). This PR rewrites it.

For context, here is the [line](https://github.com/huggingface/transformers/blob/main/src/transformers/generation_beam_search.py#L872) where it is applied: `score = sum_logprobs / (hyp.shape[-1] ** self.length_penalty)`. Since `sum_logprobs` is a negative value and `hyp.shape[-1]` is the length of the sequence (positive value), it implies that a positive `self.length_penalty` will lead to a score that is smaller in magnitude as the sequence grows = more positive = higher = this sequence has increased odds of being picked.

Finally, this means that there is a mismatch between the variable name and its effect. In practice, it is not a "penalty", as a positive value actually promotes "length". However, the alternatives are breaking changes (either changing how it is applied or the variable name), which is highly undesirable.

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for fixing!
So it means that we are always enabling length penalty by default, since 1.0 is not the neutral value, right?

**Comment by sgugger** [comment]
Merging to cherry-pick it on the release branch

**Comment by patrickvonplaten** [comment]
Thanks for the fix @gante - I think it's quite common to use a length penalty for beam search. So not sure if it's worth switching here to "no-length-penalty" by default

**Comment by zenquiorra** [comment]
@patrickvonplaten

> I believe it is quite common to employ a length penalty in beam search. Hence, I am unsure if it would be worth switching to the "no-length-penalty" option by default.

Apart from the discrepancy in terminology where "length penalty" actually refers to "length reward," let's consider the default value. In the context of beam search, the concept of length penalty implies a preference for shorter sequences. However, the current implementation seems to encourage generating longer sequences by default, as you mentioned. This contradicts the common practice of using a length penalty in beam search to prioritize shorter sequences.

Wouldn't it be more logical to set the default value to "-1" instead? This adjustment would signify that, by default, we are giving priority to short-length sequences, aligning with the expected behavior of beam search's default implementation.

**Comment by gante** [comment]
@zenquiorra 👋 

As Patrick wrote in another thread (that I can't find), slightly promoting longer sequences is often beneficial in practice. The probability of each token is `<=1.0`, which means the sequence score is expected to decrease quickly as more tokens are added. This implies that when the score of a particular sequence barely changes when more tokens are added, those additional tokens are likely to be important for the output. A small positive length penalty, promoting longer sequences, would capture this benefit.

Another important aspect is backward compatibility. While the benefits of this default value (and variable naming) are debatable, we have many projects and products built on top of `transformers`. Changing the API or default values must be done for a very strong reason (which is not the case here) 🤗
